File: GoEngineering/01_rest_api_service/explanation/01_rest_api_service_explanation.go
/*
Problem 01 — REST API Service (net/http, Go 1.22+ routing)

WHAT WE'RE BUILDING

A small but production-shaped HTTP API for managing "tasks" (think a minimal
issue tracker): create, list, fetch, update, and delete tasks, backed by an
in-memory store guarded for concurrent access. This is the shape of *every*
CRUD service you'll write in Go before a database enters the picture — get
this skeleton right and swapping the in-memory store for Postgres later is a
one-file change.

# WHY THIS MATTERS IN REAL SYSTEMS

Every backend engineer eventually owns an HTTP service. The parts that
separate "compiles and works in the demo" from "survives production" are:
  - Routing that matches on method AND path without a third-party router.
  - JSON decoding that rejects malformed/unknown input instead of silently
    zero-valuing fields.
  - Validation errors that tell the *caller* what's wrong, in a stable
    machine-readable shape — not a stack trace or a generic 500.
  - Consistent error-response envelopes so every client-facing failure looks
    the same regardless of which handler produced it.
  - A server that shuts down cleanly: stop accepting new connections, let
    in-flight requests finish, then exit — instead of dropping requests on
    SIGTERM (which is what a naive `http.ListenAndServe` does).

CONCEPTS COVERED

  - `http.ServeMux` method+path patterns (Go 1.22+): "GET /tasks/{id}"
  - `http.ResponseWriter` / `*http.Request` handler signatures
  - JSON decode with `json.Decoder.DisallowUnknownFields`
  - Structured validation with a stable error envelope
  - Sentinel errors + `errors.Is` for control flow between layers
  - `context` deadlines on the server and graceful shutdown via
    `signal.NotifyContext` + `http.Server.Shutdown`
  - `sync.RWMutex` for a concurrency-safe in-memory store

# SPEC

Resource: Task{ ID string, Title string, Done bool, CreatedAt time.Time }

Endpoints (Go 1.22 mux patterns — method + path together):

	GET    /healthz            -> 200 {"status":"ok"}
	POST   /tasks               -> create a task from JSON body {"title": "..."}
	GET    /tasks               -> list all tasks (stable order: creation order)
	GET    /tasks/{id}          -> fetch one task, 404 if missing
	PUT    /tasks/{id}          -> full update {"title": "...", "done": bool}
	DELETE /tasks/{id}          -> delete, 204 on success, 404 if missing

Validation rules:
  - Title is required, must be non-empty after trimming whitespace, and must
    be <= 200 runes.
  - Unknown JSON fields in the request body are a 400 (catches client typos
    early instead of silently ignoring them).
  - Malformed JSON is a 400 with a clear message, not a 500.

Error response envelope (ALL error responses, any status >= 400, use this
exact shape):

	{"error": {"code": "validation_failed", "message": "title is required"}}

Status code mapping:
  - validation errors            -> 400, code "validation_failed"
  - task not found                -> 404, code "not_found"
  - malformed JSON / wrong Content-Type -> 400, code "bad_request"
  - anything unexpected (panic recovered, etc.) -> 500, code "internal_error"
    (panic recovery itself is problem 02's job — here just don't let a bug
    in a handler leak a raw Go error string or stack trace to the client)

Server behavior:
  - Listens on an address supplied by the caller of NewServer.
  - Has explicit ReadHeaderTimeout / WriteTimeout / IdleTimeout set (a bare
    `http.Server{}` with no timeouts is a slow-loris footgun in production).
  - main() wires SIGINT/SIGTERM to a graceful shutdown with a bounded grace
    period: stop accepting new connections, wait for in-flight requests up
    to the grace period, then force-close.

ACCEPTANCE CRITERIA

  - `newMux` returns an `http.Handler` that can be exercised directly via
    `httptest.NewServer` or `httptest.NewRequest`/`ResponseRecorder` without
    starting a real listener — this is what the tests do.
  - All five endpoints behave per spec, including status codes and the error
    envelope shape.
  - The store is safe under concurrent GET/POST/PUT/DELETE (no data race —
    verify with `go test -race`).
  - `main()` starts a real server and shuts it down gracefully on signal.
*/
package main

import (
	"encoding/json"
	"net/http"
	"sync"
	"time"
)

// Task is the resource this API exposes.
type Task struct {
	ID        string    `json:"id"`
	Title     string    `json:"title"`
	Done      bool      `json:"done"`
	CreatedAt time.Time `json:"created_at"`
}

// TaskStore is a concurrency-safe in-memory collection of tasks.
//
// TODO: add whatever fields you need (a map keyed by ID, a mutex, an
// incrementing ID counter or a UUID generator, and something that lets List
// return tasks in stable creation order — a map alone does not preserve
// order).
type TaskStore struct {
	mu sync.RWMutex
	// TODO: fields
}

// NewTaskStore returns an empty, ready-to-use TaskStore.
func NewTaskStore() *TaskStore {
	panic("TODO: implement NewTaskStore")
}

// Create inserts a new task with the given title and returns it.
// TODO: generate an ID, set CreatedAt, store it, preserve insertion order.
func (s *TaskStore) Create(title string) Task {
	panic("TODO: implement TaskStore.Create")
}

// Get returns the task with the given id.
// TODO: return (Task{}, ErrNotFound) when the id doesn't exist.
func (s *TaskStore) Get(id string) (Task, error) {
	panic("TODO: implement TaskStore.Get")
}

// List returns all tasks in creation order.
func (s *TaskStore) List() []Task {
	panic("TODO: implement TaskStore.List")
}

// Update overwrites the title/done fields of an existing task.
// TODO: return ErrNotFound when the id doesn't exist.
func (s *TaskStore) Update(id string, title string, done bool) (Task, error) {
	panic("TODO: implement TaskStore.Update")
}

// Delete removes a task by id.
// TODO: return ErrNotFound when the id doesn't exist.
func (s *TaskStore) Delete(id string) error {
	panic("TODO: implement TaskStore.Delete")
}

// ErrNotFound is returned by TaskStore methods when an id doesn't exist.
// TODO: define this as a sentinel error (errors.New), and use errors.Is at
// the HTTP layer to translate it into a 404 without the store knowing
// anything about HTTP.
var ErrNotFound error

// apiError is the stable JSON envelope every error response uses.
//
// TODO: define the JSON shape described in the header comment:
//
//	{"error": {"code": "...", "message": "..."}}
//
// Hint: you likely want two structs — an outer one with a single "error"
// field, and an inner one with "code"/"message" — so encoding/json produces
// the nested shape without manual map-building.
type apiError struct {
	// TODO: fields
}

// writeJSON encodes v as JSON with the given status code.
// TODO: set Content-Type: application/json, write the status header, then
// encode v to w. Consider what happens if encoding fails after the header
// is already written (you can't change the status at that point — at
// minimum don't panic).
func writeJSON(w http.ResponseWriter, status int, v any) {
	panic("TODO: implement writeJSON")
}

// writeError writes an error response using the apiError envelope.
// TODO: build the envelope and delegate to writeJSON.
func writeError(w http.ResponseWriter, status int, code, message string) {
	panic("TODO: implement writeError")
}

// decodeJSON decodes the request body into dst, rejecting unknown fields
// and malformed JSON.
//
// TODO: use json.NewDecoder(r.Body), call DisallowUnknownFields, and Decode
// into dst. Also check for a trailing second JSON value in the body (a
// classic bug: `{"a":1}{"b":2}` decodes the first object silently) by
// calling Decode a second time into a throwaway `json.RawMessage` and
// expecting io.EOF.
func decodeJSON(r *http.Request, dst any) error {
	panic("TODO: implement decodeJSON")
}

// createTaskRequest is the JSON body accepted by POST /tasks.
type createTaskRequest struct {
	Title string `json:"title"`
}

// updateTaskRequest is the JSON body accepted by PUT /tasks/{id}.
type updateTaskRequest struct {
	Title string `json:"title"`
	Done  bool   `json:"done"`
}

// validateTitle enforces the title validation rules from the spec.
// TODO: trim whitespace, reject empty, reject > 200 runes. Return a
// descriptive error whose message is safe to show directly to the caller.
func validateTitle(title string) (string, error) {
	panic("TODO: implement validateTitle")
}

// handleHealthz implements GET /healthz.
func handleHealthz(w http.ResponseWriter, r *http.Request) {
	panic("TODO: implement handleHealthz")
}

// handleCreateTask implements POST /tasks.
// TODO: decode+validate the body, call store.Create, respond 201 with the
// created task.
func handleCreateTask(store *TaskStore) http.HandlerFunc {
	panic("TODO: implement handleCreateTask")
}

// handleListTasks implements GET /tasks.
func handleListTasks(store *TaskStore) http.HandlerFunc {
	panic("TODO: implement handleListTasks")
}

// handleGetTask implements GET /tasks/{id}.
// TODO: read the {id} path value via r.PathValue("id") (Go 1.22+ mux
// feature — no third-party router needed).
func handleGetTask(store *TaskStore) http.HandlerFunc {
	panic("TODO: implement handleGetTask")
}

// handleUpdateTask implements PUT /tasks/{id}.
func handleUpdateTask(store *TaskStore) http.HandlerFunc {
	panic("TODO: implement handleUpdateTask")
}

// handleDeleteTask implements DELETE /tasks/{id}.
// TODO: 204 No Content on success — no body.
func handleDeleteTask(store *TaskStore) http.HandlerFunc {
	panic("TODO: implement handleDeleteTask")
}

// newMux wires every route to its handler and returns the resulting
// http.Handler. Tests exercise the server through this function directly,
// without a real network listener.
//
// TODO: build an *http.ServeMux, register each route using Go 1.22+
// "METHOD /path" patterns, and return it.
func newMux(store *TaskStore) http.Handler {
	panic("TODO: implement newMux")
}

// newServer builds an *http.Server with production-sane timeouts wrapping
// the mux from newMux.
// TODO: set Addr, Handler, ReadHeaderTimeout, WriteTimeout, IdleTimeout.
func newServer(addr string, store *TaskStore) *http.Server {
	panic("TODO: implement newServer")
}

func main() {
	// TODO:
	//  1. Build a store and server via newServer(":8080", NewTaskStore()).
	//  2. Start it in a goroutine with ListenAndServe, sending any non-
	//     ErrServerClosed error to a channel (or log.Fatal-ing carefully —
	//     but prefer not crashing the whole process from a goroutine you
	//     can't recover from cleanly).
	//  3. Wait for SIGINT/SIGTERM via signal.NotifyContext.
	//  4. On signal, call server.Shutdown(ctx) with a bounded timeout
	//     (e.g. 10s) so in-flight requests can finish, then exit.
	panic("TODO: implement main")
}

var _ = json.Marshal // keep encoding/json imported for the stub file

/*
HINTS

  - `r.PathValue("id")` is the Go 1.22+ way to read a `{id}` path segment —
    no third-party router required. It only works if the mux pattern
    actually contains `{id}`.
  - `http.ServeMux` patterns that specify a method ("GET /tasks") only match
    that method; an unmatched method on a registered path returns 405
    automatically as of Go 1.22 — you don't have to check r.Method yourself
    inside the handler.
  - Prefer building the mux as a plain function (`newMux`) rather than a
    package-level `http.DefaultServeMux` + `init()` — package-level mutable
    routing state makes tests fight each other.
  - `json.Decoder.DisallowUnknownFields` + a second `Decode` call to detect
    trailing garbage are both real production patterns, not academic — a
    client sending `{"title":"x"}{"evil":1}` or a typo'd field name should
    fail loudly, not silently.

COMMON PITFALLS

  - Returning raw Go errors (`err.Error()`) to the client — leaks internal
    detail and is inconsistent in shape. Always go through writeError.
  - Forgetting `WriteHeader` must be called before `Write`/`Encode` writes
    body bytes — encoding/json's Encode after WriteHeader is fine, but if
    you write body bytes first, the status silently becomes 200.
  - A `TaskStore` using a `map[string]Task` with no separate ordering slice
    — Go map iteration order is randomized, so `List` will return tasks in
    a different order every call. Track insertion order explicitly.
  - `http.ListenAndServe` blocking main forever with no signal handling —
    means `docker stop` / k8s SIGTERM hard-kills the process after its
    grace period, dropping in-flight requests.
  - Validating only in the handler and not returning descriptive messages —
    "400 Bad Request" with no body forces the API consumer to guess.

STRETCH GOALS

  - Add pagination to GET /tasks (`?limit=&cursor=`).
  - Add optimistic concurrency: an ETag/If-Match header on PUT, 412 on
    mismatch.
  - Add a PATCH /tasks/{id} for partial updates (JSON merge semantics —
    distinguish "field omitted" from "field set to zero value" using
    pointer fields or `json.RawMessage`).
  - Swap TaskStore's backing storage for a `database/sql` implementation
    behind the same interface (foreshadows problem 09).
*/

File: PyEngineering/01_rest_api_service/01_rest_api_service_explanation.py
"""
01 · REST API service
=====================

WHAT WE'RE BUILDING
--------------------
```arch
%% caption: REST API Service Architecture
node req "Client Request" at 0,1 icon=client color=blue
node api "FastAPI Router\n(HTTP endpoints)" at 1,1 icon=api color=green
node valid "Pydantic\n(Validation)" at 2,1 icon=check color=amber
node domain "Domain Logic\n(CRUD ops)" at 3,1 icon=process color=slate
node db "In-Memory Store\n(with asyncio.Lock)" at 4,1 icon=memory color=red

req -> api
api -> valid : "parses JSON"
valid -> domain : "valid payload"
domain -> db : "safe mutation"
```

A small but production-shaped "Tasks" REST API on FastAPI: full CRUD over an
in-memory store, Pydantic request/response validation, a typed custom
exception hierarchy mapped to HTTP status codes, and a `lifespan` context
manager that models graceful startup/shutdown — the same shape you'd use to
open/close a DB pool, an HTTP client, or a message-bus connection in a real
service.

WHY THIS MATTERS IN REAL SYSTEMS
---------------------------------
Almost every backend Python service written today is "a FastAPI app with a
handful of routers." The parts that separate a toy example from something
you'd put in front of traffic are exactly the parts this problem drills:

- **Validation at the boundary.** Pydantic models mean malformed input never
  reaches your business logic — FastAPI returns a 422 automatically, with a
  machine-readable error body, before your code runs at all.
- **A typed error taxonomy.** Raising bare `HTTPException`s everywhere
  couples your domain logic to HTTP. A domain exception
  (`TaskNotFoundError`) raised deep in a store method, caught once by an
  `@app.exception_handler`, keeps the store layer HTTP-agnostic and testable
  on its own.
- **Concurrency-safety of shared state.** FastAPI runs each request handler
  as its own task on one event loop. An in-memory dict "store" is *not*
  automatically safe for interleaved read-modify-write sequences across
  concurrent requests unless you protect the critical section — this is the
  single most common bug when people hand-roll an in-memory store under
  async handlers.
- **Graceful shutdown.** A `lifespan` context manager is where you open
  resources once (not per-request) and guarantee they're closed even if the
  process receives SIGTERM (as it will, constantly, in a container
  orchestrator doing a rolling deploy).

CONCEPTS COVERED
-----------------
- FastAPI routing (`APIRouter`, path/query params, response models)
- Pydantic v2 models: validation, `model_config`, computed/default fields
- Custom exception hierarchy + `@app.exception_handler`
- `asyncio.Lock` to guard shared in-memory state under concurrent handlers
- `lifespan` context manager (the modern replacement for
  `@app.on_event("startup"/"shutdown")`)
- Graceful shutdown story: what `lifespan` guarantees vs. what
  `uvicorn`'s SIGTERM handling does at the process level (explained in
  comments; not exercised by the test suite, which drives the app via
  `TestClient`/`httpx.ASGITransport` rather than a live server+signals)

THE SPEC
--------
Build a `Task` resource:

    Task:
        id:         str   (server-generated UUID4 hex, read-only)
        title:      str   (1..200 chars, required)
        done:       bool  (default False)
        created_at: datetime (server-generated, UTC, read-only)

Endpoints (all under `/tasks`):

    POST   /tasks          create a task from {title, done?}      -> 201, Task
    GET    /tasks          list tasks, optional ?done=true/false  -> 200, list[Task]
    GET    /tasks/{id}     fetch one task                         -> 200, Task | 404
    PATCH  /tasks/{id}     partial update {title?, done?}         -> 200, Task | 404
    DELETE /tasks/{id}     delete a task                          -> 204 | 404

Validation rules:
    - `title` must be 1..200 characters after stripping whitespace; a
      title that is empty/whitespace-only is a 422, not a 400 — that's
      FastAPI/Pydantic's job, not yours to hand-code.
    - `PATCH` body fields are all optional; omitted fields are left
      unchanged. At least one field must be present, or respond 400 (this
      one *is* your job — Pydantic can't express "not all fields empty" for
      a partial-update body without extra work not worth it here, so it's a
      manual check in the handler).

Error mapping:
    - `TaskNotFoundError` (raised by the store)  -> 404 with
      `{"detail": "task <id> not found"}`
    - Pydantic validation failure                -> 422 (FastAPI's default,
      do not override it)

ACCEPTANCE CRITERIA
--------------------
1. `from NN_rest_api_service_solution import app` succeeds with no side
   effects (no server starts on import — `app` is just the ASGI app object).
2. All five endpoints behave per the table above, verified via
   `fastapi.testclient.TestClient` (which drives the ASGI app in-process,
   including running `lifespan` startup/shutdown — no real socket needed).
3. Two concurrent requests that both mutate the store (e.g. two concurrent
   `POST`s) never corrupt the store's internal dict — protected by an
   `asyncio.Lock`.
4. `lifespan` startup populates `app.state` with the store and logs a
   startup message; shutdown clears the store and logs a shutdown message —
   both observable in a test using `with TestClient(app) as client:`.
5. mypy-clean, black-formatted, ruff-clean.
"""

from __future__ import annotations

from collections.abc import AsyncIterator
from contextlib import asynccontextmanager

from fastapi import FastAPI
from pydantic import BaseModel

# NOTE (stub only): you'll need more imports once you fill in the TODOs
# below — at minimum `uuid` (task ids), `datetime.UTC`/`datetime.datetime`
# (timestamps), `pydantic.Field` (length-constrained fields), and from
# `fastapi`: `HTTPException`, `Query` (the `?done=` filter), `Request`
# (reaching `request.app.state.store`), and `status` (named status codes).
# They're listed here rather than pre-imported so this file imports cleanly
# as a stub without tripping an unused-import lint error.


# ---------------------------------------------------------------------------
# Domain exceptions — HTTP-agnostic. The store layer should never import
# `fastapi` or raise `HTTPException` directly; that keeps it testable and
# reusable outside of a web context (e.g. from a CLI or a background job).
# ---------------------------------------------------------------------------
class TaskError(Exception):
    """Base class for all task-store domain errors."""


class TaskNotFoundError(TaskError):
    """Raised when a task id does not exist in the store."""

    def __init__(self, task_id: str) -> None:
        # TODO: store `task_id` on the instance and call
        # super().__init__ with a human-readable message. Callers (the
        # exception handler) will read the message via str(exc).
        raise NotImplementedError("TODO: implement TaskNotFoundError.__init__")


# ---------------------------------------------------------------------------
# Pydantic models — the request/response contract.
# ---------------------------------------------------------------------------
class TaskCreate(BaseModel):
    """Body for POST /tasks."""

    # TODO: `title: str` constrained to 1..200 chars after stripping
    # whitespace (use pydantic.Field with min_length/max_length, and either
    # a field validator that strips, or str_strip_whitespace in model_config).
    # TODO: `done: bool = False`


class TaskUpdate(BaseModel):
    """Body for PATCH /tasks/{id}. Every field optional (partial update)."""

    # TODO: `title: str | None = None` (same length constraint when present)
    # TODO: `done: bool | None = None`


class TaskOut(BaseModel):
    """Response model. Server-generated fields are never client-writable."""

    # TODO: id: str, title: str, done: bool, created_at: datetime


# ---------------------------------------------------------------------------
# Store — an async-safe in-memory repository. In a real service this would
# be a thin wrapper over a DB connection pool; the concurrency-safety shape
# (guard the critical section, keep methods small) is identical either way.
# ---------------------------------------------------------------------------
class TaskStore:
    """In-memory task repository, safe under concurrent async handlers."""

    def __init__(self) -> None:
        # TODO: self._tasks: dict[str, TaskOut] = {}
        # TODO: self._lock = asyncio.Lock()
        raise NotImplementedError("TODO: implement TaskStore.__init__")

    async def create(self, data: TaskCreate) -> TaskOut:
        """Generate an id + created_at, store, and return the new task."""
        # TODO: build a TaskOut with uuid.uuid4().hex and datetime.now(UTC),
        # insert it under the lock, return it.
        raise NotImplementedError("TODO: implement TaskStore.create")

    async def list(self, *, done: bool | None = None) -> list[TaskOut]:
        """Return all tasks, optionally filtered by `done`."""
        raise NotImplementedError("TODO: implement TaskStore.list")

    async def get(self, task_id: str) -> TaskOut:
        """Return one task or raise TaskNotFoundError."""
        raise NotImplementedError("TODO: implement TaskStore.get")

    async def update(self, task_id: str, data: TaskUpdate) -> TaskOut:
        """Apply only the fields set on `data`; raise TaskNotFoundError if missing."""
        raise NotImplementedError("TODO: implement TaskStore.update")

    async def delete(self, task_id: str) -> None:
        """Remove a task or raise TaskNotFoundError."""
        raise NotImplementedError("TODO: implement TaskStore.delete")


# ---------------------------------------------------------------------------
# Lifespan — startup/shutdown. This is where you'd open a DB pool or an
# httpx.AsyncClient once per process and guarantee it's closed on shutdown,
# including on SIGTERM (uvicorn translates SIGTERM into a clean ASGI
# lifespan "shutdown" event, which is what makes this pattern "graceful").
# ---------------------------------------------------------------------------
@asynccontextmanager
async def lifespan(app: FastAPI) -> AsyncIterator[None]:
    # TODO: create a TaskStore, assign it to app.state.store, log/print a
    # startup message, `yield`, then log/print a shutdown message and clear
    # the store. Everything before `yield` runs on startup; everything after
    # runs on shutdown — even if the process is asked to stop while requests
    # are in flight, ASGI servers drain in-flight requests before calling
    # the shutdown half of lifespan.
    raise NotImplementedError("TODO: implement lifespan")
    yield  # pragma: no cover - unreachable, keeps this a valid generator


def create_app() -> FastAPI:
    """Application factory — lets tests build fresh app instances."""
    app = FastAPI(lifespan=lifespan)

    # TODO: register routes on `app` (either directly or via an APIRouter),
    # and register an exception_handler for TaskNotFoundError that returns
    # a 404 JSONResponse. Access the store as request.app.state.store inside
    # handlers, never as a module-level global — that's what makes the app
    # testable with multiple independent instances.
    raise NotImplementedError("TODO: implement create_app routes")

    return app  # pragma: no cover - unreachable until TODO above is done


# NOTE (stub only): unlike the solution file, `app` is *not* built at import
# time here. `create_app()` still raises `NotImplementedError` until you fill
# in the TODOs above, and this file must import cleanly as-is — so calling
# it unconditionally at module scope would break that contract. Once your
# implementation is in place, either uncomment the line below or drive
# `create_app()` from your own `if __name__ == "__main__":` block:
#
#     app = create_app()


# ---------------------------------------------------------------------------
# HINTS
# -----
# - `request.app.state.store` is the idiomatic way to reach per-app state
#   from inside a route handler without module-level globals.
# - Prefer `response_model=TaskOut` (or a return-type annotation FastAPI can
#   introspect) over building JSONResponse by hand — you get automatic
#   OpenAPI docs and response validation for free.
# - `TestClient` from `fastapi.testclient` runs the full ASGI lifecycle
#   (including `lifespan`) when used as a context manager:
#   `with TestClient(app) as client: ...` — a bare `TestClient(app)` without
#   the `with` does NOT run lifespan in recent FastAPI/Starlette versions.
# - `HTTPException(status_code=..., detail=...)` is fine for hand-authored
#   4xx paths inside a handler (e.g. the "at least one field" 400 on
#   PATCH); reserve the domain exception hierarchy for errors raised deep
#   in the store, away from any `Request`/`HTTPException` import.
#
# COMMON PITFALLS
# ---------------
# - Forgetting the `asyncio.Lock` and doing "check-then-act" on the dict
#   (e.g. `if id in self._tasks: del self._tasks[id]`) — under concurrent
#   requests two coroutines can both pass the check before either acts.
#   The lock must wrap the *entire* read-modify-write sequence, not just
#   the mutation.
# - Returning `204 No Content` with a body — FastAPI/Starlette will reject
#   this; a 204 response must have an empty body (don't set
#   `response_model` on the DELETE handler).
# - Using a *module-level* `TaskStore()` instance instead of `app.state` —
#   this leaks state across test cases that each construct their own `app`
#   and makes lifespan startup/shutdown meaningless (there's nothing left
#   for it to create/destroy).
# - `datetime.utcnow()` is deprecated in modern Python — use
#   `datetime.now(UTC)` for a timezone-aware timestamp.
#
# STRETCH GOALS
# --------------
# - Add pagination (`?limit=&cursor=`) to `GET /tasks` instead of returning
#   the whole store.
# - Add an `ETag`/`If-Match` optimistic-concurrency check on `PATCH`.
# - Swap the in-memory dict for stdlib `sqlite3` (foreshadows problem 09)
#   without changing the router code — only `TaskStore`'s internals change,
#   which is the point of keeping HTTP and storage layered.
# - Add a real uvicorn entrypoint (`if __name__ == "__main__":
#   uvicorn.run(...)`) and manually send SIGTERM to a running process to
#   observe the lifespan shutdown log fire before the process exits.
